src/process_utility.c

```diff
@@ -600,6 +600,70 @@ add_chunk_to_vacuum(Hypertable *ht, Oid chunk_relid, void *arg)
 	ctx->chunk_rels = lappend(ctx->chunk_rels, chunk_vacuum_rel);
 }
 
+/*
+ * Construct a list of VacuumRelations for all vacuumable rels in
+ * the current database.  This is similar to the PostgresQL get_all_vacuum_rels
+ * from vacuum.c, only it filters out distributed hypertables and chunks
+ * that have been compressed.
+ */
+static List *
+ts_get_all_vacuum_rels(bool is_vacuumcmd)
+{
+	List *vacrels = NIL;
+	Relation pgclass;
+	TableScanDesc scan;
+	HeapTuple tuple;
+	Cache *hcache = ts_hypertable_cache_pin();
+
+	pgclass = table_open(RelationRelationId, AccessShareLock);
+
+	scan = table_beginscan_catalog(pgclass, 0, NULL);
+
+	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	{
+		Form_pg_class classForm = (Form_pg_class) GETSTRUCT(tuple);
+		Oid relid = classForm->oid;
+		Hypertable *ht;
+		Chunk *chunk;
+
+		/* check permissions of relation */
+		if (!vacuum_is_relation_owner(relid,
+									  classForm,
+									  is_vacuumcmd ? VACOPT_VACUUM : VACOPT_ANALYZE))
+			continue;
+
+		/*
+		 * We include partitioned tables here; depending on which operation is
+		 * to be performed, caller will decide whether to process or ignore
+		 * them.
+		 */
+		if (classForm->relkind != RELKIND_RELATION && classForm->relkind != RELKIND_MATVIEW &&
+			classForm->relkind != RELKIND_PARTITIONED_TABLE)
+			continue;
+
+		ht = ts_hypertable_cache_get_entry(hcache, relid, CACHE_FLAG_MISSING_OK);
+		if (ht && hypertable_is_distributed(ht))
+			continue;
+
+		chunk = ts_chunk_get_by_relid(relid, false);
+		if (chunk && chunk->fd.compressed_chunk_id != INVALID_CHUNK_ID)
+			continue;
```

Comment:
If a hypertable was found, no need to check for chunk:

```suggestion
  	ht = ts_hypertable_cache_get_entry(hcache, relid, CACHE_FLAG_MISSING_OK);
		
		if (ht)
		{
		    if (hypertable_is_distributed(ht))
			    continue;
	    }
	    else
	    {
           Chunk *chunk = ts_chunk_get_by_relid(relid, false);
             
		    if (chunk && chunk->fd.compressed_chunk_id != INVALID_CHUNK_ID)
			   continue;
	     }
```

Reply:
fixed